### PR TITLE
seastar: update submodule 

### PR DESCRIPTION
**Pull request by Matan-B**
* #58427 reverted a commit in the submodule in order to resolve a FTBFS which was fixed in latest seastar https://github.com/ceph/seastar/commit/378b776050edded86da23b1e753d775c3b27bc7d.
The revert is removed.

*  `Seastar_STD_OPTIONAL_VARIANT_STRINGVIEW` is removed

* `Seastar_DEPRECATED_OSTREAM_FORMATTERS OFF` - set to OFF to use seastar's formatters.

* Latest seastar deprecated get0 - use get instead

* WIP: clang 14 FTBFS - https://github.com/ceph/seastar/commit/e7a2679235aeea4c9635e5202260204f0147a8c2

---

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by Matan-B** [comment]
jenkins test make check

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
@tchaikov, @athanatos - I had to also revert https://github.com/ceph/seastar/commit/914a4241c36c41a327cfbe77813bffc6d0de1274 from our submodule since we can't deprecate operator<< yet.  The commit message says:
```
and these two formatters are marked deprecated.
so in future, we can remove them when we bump up the API level or just
completely drop them.
```
Is it intended for the deprecation to apply to the project as a whole (and not just seastar's usage of it)?

**Comment by tchaikov** [comment]
> @tchaikov, @athanatos - I had to also revert [ceph/seastar@914a424](https://github.com/ceph/seastar/commit/914a4241c36c41a327cfbe77813bffc6d0de1274) from our submodule since we can't deprecate operator<< yet (for vector and unordered_map). The commit message says:
> 
> ```
> ...
> and these two formatters are marked deprecated.
> so in future, we can remove them when we bump up the API level or just
> completely drop them.
> ```
> 
> Is it intended for the deprecation to apply to the project as a whole (and not just seastar's usage of it)?

IMO, seastar is just an event-driven library, it cannot decide how its parent project format an instance of `std::vector<T>`. actually, ceph has two overloads for formatting `std::vector<T>`, and one overload for formatting `std::unordered_set<T>`.

i think the right way to fix this is

1. do not define `Seastar_DEPRECATED_OSTREAM_FORMATTERS`
2. define `operator<<` for `std::unordered_map<T>` in Ceph, instead of relying on a library like seastar to do this.

**Comment by Matan-B** [comment]
> i think the right way to fix this is
> 
> 1. do not define `Seastar_DEPRECATED_OSTREAM_FORMATTERS`
> 2. define `operator<<` for `std::unordered_map<T>` in Ceph, instead of relying on a library like seastar to do this.

Thank you for the quick reply.
btw, do you think that https://github.com/ceph/seastar/commit/e7a2679235aeea4c9635e5202260204f0147a8c2 should be pushed to the main repo as well? Awaiting make check to find more instances.

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-seastar-july7

**Comment by Matan-B** [comment]
Added https://tracker.ceph.com/issues/66905 to remove the last commit from our submodule once we update the compilers used for CI.

**Comment by tchaikov** [comment]
> > i think the right way to fix this is
> > 
> > 1. do not define `Seastar_DEPRECATED_OSTREAM_FORMATTERS`
> > 2. define `operator<<` for `std::unordered_map<T>` in Ceph, instead of relying on a library like seastar to do this.
> 
> @tchaikov, thank you for the quick reply. btw, do you think that [ceph/seastar@e7a2679](https://github.com/ceph/seastar/commit/e7a2679235aeea4c9635e5202260204f0147a8c2) should be pushed to the main repo as well? Awaiting make check to find more instances.

unless you believe it's a bug in seastar instead of a bug in clang.

you could verify this with, for instance, a recent version of GCC. if both of them complain, there are good chances it's a bug in Seastar. or you could check if this is indeed something not supported by C++20. yeah, i knew that the standard is not human readable under most circumstances. but still..

> 
> Looks clang specific [llvm/llvm-project#52720](https://github.com/llvm/llvm-project/issues/52720)
> 
> Edit: structural bindings captures were is the issue for clang 15 and below. https://godbolt.org/z/ez4Y99xfK

**Comment by Matan-B** [comment]
> unless you believe it's a bug in seastar instead of a bug in clang.
 
It's a bug in clang which was fixed in 16 and higher, shouldn't probably be fixed in seastar. I'll keep it in out submodule until we stop using the older version.

**Review by athanatos — APPROVED** [review]

File: src/test/crimson/seastore/test_seastore.cc [hunk 1]
```diff
@@ -83,15 +83,15 @@ struct seastore_test_t :
   }
 
   void do_transaction(CTransaction &&t) {
-    return sharded_seastore->do_transaction(
+    return (void)sharded_seastore->do_transaction(
```
Comment (tchaikov) [inline src/test/crimson/seastore/test_seastore.cc:86]: nit, s/return//

**Comment by cbodley** [comment]
@Matan-B can this be reasonably backported to squid, or should we backport https://github.com/ceph/ceph/pull/58427 instead? still seeing a lot of 'make check' failures on squid due to https://tracker.ceph.com/issues/66834

**Comment by Matan-B** [comment]
> @Matan-B can this be reasonably backported to squid, or should we backport #58427 instead? still seeing a lot of 'make check' failures on squid due to https://tracker.ceph.com/issues/66834

https://github.com/ceph/ceph/pull/58702 should do. thanks for updating the tracker!

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-07-11_12:17:07-crimson-rados-wip-matanb-seastar-july7-distro-crimson-smithi/

**Comment by Matan-B** [comment]
@xxhdx1985126 @cyx1231st FYI

**Pull request merged**